### PR TITLE
Create test to cover path of azureMachinePool.Spec.UserAssignedIdenti…

### PR DESCRIPTION
**Pull request by troy0820**
**What type of PR is this?**
/kind cleanup 

**What this PR does / why we need it**:
Adds testing to the `controllers` package 

- [x] Add unit test for controller azurejson
- [x] Create func to monkey patch during testing
- [x] Export interface instead of concrete type for testing 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/4277
Partially fixes: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3649

**Special notes for your reviewer**:

_I had to change the return value to an interface _(which the return type implements)_ to allow the patch to ensure mocking can take place.  The mocking will use the generated mocks and the EXPECT() to allow to hit that code path while testing._ 

- [ ] cherry-pick candidate

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: azure/services/identities/client.go [hunk 1]
```diff
@@ -39,7 +39,7 @@ type AzureClient struct {
 }
 
 // NewClient creates a new MSI client from an authorizer.
-func NewClient(auth azure.Authorizer) (*AzureClient, error) {
+func NewClient(auth azure.Authorizer) (Client, error) {
```
Comment (troy0820) [inline azure/services/identities/client.go:42]: This doesn't change anything but allow this interface to be mocked as a return value for testing.  The concrete type is still in use here as the return value for that is the concrete type which implements this interface.

File: controllers/azurejson_machinepool_controller.go [hunk 2]
```diff
@@ -194,3 +194,7 @@ func (r *AzureJSONMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl
 
 	return ctrl.Result{}, nil
 }
+
+var getClient = func(auth azure.Authorizer) (identities.Client, error) {
```
Comment (troy0820) [inline controllers/azurejson_machinepool_controller.go:198]: This "patch" allows for this to be overridden during testing but not in production.

File: controllers/azurejson_machinepool_controller_test.go [hunk 3]
```diff
@@ -18,17 +18,24 @@ package controllers
 
 import (
 	"context"
-	"os"
```
Comment (troy0820) [inline controllers/azurejson_machinepool_controller_test.go:21]: Removed to use `t.Setenv`

File: controllers/azurejson_machinepool_controller.go [hunk 4]
```diff
@@ -114,7 +114,6 @@ func (r *AzureJSONMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl
 		}
 		return reconcile.Result{}, err
 	}
-
```
Comment (willie-yao) [inline controllers/azurejson_machinepool_controller.go:117]: nit: You probably didn't mean to make this change

File: controllers/azurejson_machinepool_controller_test.go [hunk 5]
```diff
@@ -210,3 +217,140 @@ func TestAzureJSONPoolReconciler(t *testing.T) {
...
 	}
 }
+
+func TestAzureJSONPoolReconcilerUserAssignedIdentities(t *testing.T) {
+	g := NewWithT(t)
+	ctrlr := gomock.NewController(t)
+	defer ctrlr.Finish()
+	req := ctrl.Request{NamespacedName: types.NamespacedName{Name: "fake-machine-pool", Namespace: "fake-ns"}}
+	ctx := context.Background()
+	scheme := runtime.NewScheme()
```
Comment (willie-yao) [inline controllers/azurejson_machinepool_controller_test.go:227]: Similar to the other PR, I think you can use this existing function: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/azurejson_machine_controller_test.go#L240

File: controllers/azurejson_machinepool_controller_test.go [hunk 6]
```diff
@@ -210,3 +217,140 @@ func TestAzureJSONPoolReconciler(t *testing.T) {
...
+		Spec: infrav1exp.AzureMachinePoolSpec{
+			UserAssignedIdentities: []infrav1.UserAssignedIdentity{
+				{
+					ProviderID: "fake-id",
+				},
+			},
+		},
+	}
+
+	clusterA := &clusterv1.Cluster{
```
Comment (willie-yao) [inline controllers/azurejson_machinepool_controller_test.go:269]: nit: I think naming this just `cluster` would be better since the "A" implies there's another cluster (or at least that's how I see it)

File: controllers/azurejson_machinepool_controller_test.go [hunk 7]
```diff
@@ -210,3 +217,140 @@ func TestAzureJSONPoolReconciler(t *testing.T) {
...
+	azureMP := &infrav1exp.AzureMachinePool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-machine-pool",
+			Namespace: "fake-ns",
+			Labels: map[string]string{
+				clusterv1.ClusterNameLabel: "fake-cluster",
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: expv1.GroupVersion.Group + "/" + expv1.GroupVersion.Version,
```
Comment (willie-yao) [inline controllers/azurejson_machinepool_controller_test.go:253]: nit: I think the convention in tests is to use fmt.Sprintf for concats. Example: https://github.com/willie-yao/cluster-api-provider-azure/blob/aks-fleet/api/v1beta1/azuremachine_validation_test.go#L201

File: controllers/azurejson_machinepool_controller_test.go [hunk 8]
```diff
@@ -210,3 +217,140 @@ func TestAzureJSONPoolReconciler(t *testing.T) {
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-other-machine-pool",
+			Namespace: "fake-ns",
+			Labels: map[string]string{
+				clusterv1.ClusterNameLabel: "fake-cluster",
+			},
+		},
+	}
+
+	azureCluster := &infrav1.AzureCluster{
```
Comment (willie-yao) [inline controllers/azurejson_machinepool_controller_test.go:293]: I think you can use this `getFakeAzureCluster` function that exists here: https://github.com/willie-yao/cluster-api-provider-azure/blob/aks-fleet/controllers/azuremachine_controller_test.go#L529

However, I think there could be a case to be made to create a new file that contains all of these functions for tests, since it's a bit confusing they exist in specific controller tests but can be used across the package. I think this could be done in a follow-up PR since it'll be a bit out of scope for this issue.

File: controllers/azurejson_machinepool_controller_test.go [hunk 9]
```diff
@@ -210,3 +217,140 @@ func TestAzureJSONPoolReconciler(t *testing.T) {
...
+	}
+
+	client := fake.NewClientBuilder().WithScheme(scheme).WithObjects(azureMP, ownerMP, clusterA, azureCluster, sec).Build()
+	rec := AzureJSONMachinePoolReconciler{
+		Client:           client,
+		Recorder:         record.NewFakeRecorder(42),
+		ReconcileTimeout: 20 * time.Second,
+	}
+	id := "fake-id"
+	getClient = func(auth azure.Authorizer) (identities.Client, error) {
```
Comment (willie-yao) [inline controllers/azurejson_machinepool_controller_test.go:348]: Where is this function used?

File: controllers/azurejson_machinepool_controller_test.go [hunk 10]
```diff
@@ -210,3 +217,140 @@ func TestAzureJSONPoolReconciler(t *testing.T) {
...
+				},
+			},
+		},
+	}
+
+	client := fake.NewClientBuilder().WithScheme(scheme).WithObjects(azureMP, ownerMP, clusterA, azureCluster, sec).Build()
+	rec := AzureJSONMachinePoolReconciler{
+		Client:           client,
+		Recorder:         record.NewFakeRecorder(42),
+		ReconcileTimeout: 20 * time.Second,
```
Comment (willie-yao) [inline controllers/azurejson_machinepool_controller_test.go:345]: I think we can use`reconciler.DefaultLoopTimeout` here rather than hard-coding the value

File: controllers/azurejson_machinepool_controller_test.go (hunk 6 above)
Comment (troy0820) [reply controllers/azurejson_machinepool_controller_test.go:269]: I didn't want to shadow any variable but I can change this.

File: controllers/azurejson_machinepool_controller_test.go (hunk 9 above)
Comment (troy0820) [reply controllers/azurejson_machinepool_controller_test.go:348]: `getClient` equals `identities.NewClient` 

If we don't patch that we have to pass an authorizer into the `identities.NewClient` func and try to auth to Azure.  Even if we pass an mocked authorizer, it will still try to go out and make the call to Aure to provide the "identity" client.  

This allows us to override this method during testing but not during production.  

In the test we do this 
```go
 getClient = func(auth azure.Authorizer) (identities.Client, error) {
       mockClient := mock_identities.NewMockClient(ctrlr)
      mockClient.EXPECT().GetClientID(gomock.Any(), gomock.Any()).Return(id, nil)
      return mockClient, nil
     }
```
So when we call `Reconcile, it doesn't try to go out and make a call to Azure but use our mock_identities caller in this method.`

So basically "patching" the method so we can test or else we get blocked here from a unit perspective.

File: controllers/azurejson_machinepool_controller_test.go (hunk 6 above)
Comment (troy0820) [reply controllers/azurejson_machinepool_controller_test.go:269]: It doesn't shadow, so I'll change this.

File: controllers/azurejson_machinepool_controller_test.go (hunk 8 above)
Comment (troy0820) [reply controllers/azurejson_machinepool_controller_test.go:293]: I would definitely do this but to migrate from this to that may be additional effort that may pollute the addition of this test.  

Also, setting up fixtures to settle all cases where that may be useful may also pollute the repo where that may need to be tested to ensure code coverage (I know that metric isn't bound to quality, but I digress).  

I can create an issue and follow up to call that out as something we would like to have.

File: controllers/azurejson_machinepool_controller_test.go (hunk 10 above)
Comment (troy0820) [reply controllers/azurejson_machinepool_controller_test.go]: Updated on latest push, resolving.

File: controllers/azurejson_machinepool_controller_test.go (hunk 7 above)
Comment (troy0820) [reply controllers/azurejson_machinepool_controller_test.go]: Utilized `fmt.Sprintf()` convention as suggested.  Resolving.

File: controllers/azurejson_machinepool_controller_test.go (hunk 6 above)
Comment (troy0820) [reply controllers/azurejson_machinepool_controller_test.go]: Changed to cluster as suggested, resolving.

File: controllers/azurejson_machinepool_controller_test.go (hunk 5 above)
Comment (troy0820) [reply controllers/azurejson_machinepool_controller_test.go]: Utilized the `newScheme` test method to grab all the schemes as suggested.  Resolving.

File: controllers/azurejson_machinepool_controller.go (hunk 4 above)
Comment (troy0820) [reply controllers/azurejson_machinepool_controller.go]: Updated this to include the space unintended for this change.  Resolving.

File: controllers/azurejson_machinepool_controller_test.go (hunk 8 above)
Comment (willie-yao) [reply controllers/azurejson_machinepool_controller_test.go:280]: Sounds good! I think it would be good to leave as-is for the current PR

**Review by willie-yao — APPROVED** [review]
/lgtm

Thanks for your awesome work on this!

**Review by CecileRobertMichon — COMMENTED** [review]
/approve 

Welcome @troy0820 🥳
